From 0f689cc00f47f93bc139c8caf3aa1ecab04a069b Mon Sep 17 00:00:00 2001
From: Thiemo van Engelen <tvanengelen@victronenergy.com>
Date: Mon, 19 Jul 2021 11:15:19 +0200
Subject: [PATCH 2/2] Wait for start/stopDiscovery call to finish

The code would send a startDiscovery or stopDiscovery call over the dbus
without waiting for the response. This could lead to wrong behaviour in
Bluez where it would still perform a discovery even after the call to
stopDiscovery due to slow processing of the startDiscovery call. By
waiting for the response of the startDiscovery, we can be sure that it
is fully processed in Bluez and that a stopDiscovery will do what it is
supposed to do.

This is usefull for Venus system that are used as CI runner for our BLE
products. It is not used in regular Venus functionality.
Upstream-Status: Inappropriate [Venus specific]
---
 src/bluetooth/bluez/bluez5_helper.cpp | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/bluetooth/bluez/bluez5_helper.cpp b/src/bluetooth/bluez/bluez5_helper.cpp
index 3b74ee64..61f84252 100644
--- a/src/bluetooth/bluez/bluez5_helper.cpp
+++ b/src/bluetooth/bluez/bluez5_helper.cpp
@@ -347,8 +347,11 @@ bool QtBluezDiscoveryManager::registerDiscoveryInterest(const QString &adapterPa
 
     d->references[adapterPath] = data;
 
-    if (!data->wasListeningAlready)
-        iface.StartDiscovery();
+    if (!data->wasListeningAlready) {
+        auto reply = iface.StartDiscovery();
+        reply.waitForFinished();
+        return !reply.isError();
+    }
 
     return true;
 }
@@ -368,7 +371,8 @@ void QtBluezDiscoveryManager::unregisterDiscoveryInterest(const QString &adapter
     if (!data->wasListeningAlready) { // Qt turned discovery mode on, Qt has to turn it off again
         OrgBluezAdapter1Interface iface(QStringLiteral("org.bluez"), adapterPath,
                                         QDBusConnection::systemBus());
-        iface.StopDiscovery();
+        auto reply = iface.StopDiscovery();
+        reply.waitForFinished();
     }
 
     delete data->propteryListener;
-- 
2.40.0.windows.1

